src/conn/conn_layered.c

```diff
@@ -779,6 +827,98 @@ __disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT
       "Updating prune timestamp failed");
 
+err:
+    return (ret);
+}
+
+/*
+ * __disagg_pick_up_checkpoint --
+ *     Pick up a new checkpoint.
+ */
+static int
+__disagg_pick_up_checkpoint(WT_SESSION_IMPL *session, const WT_DISAGG_CHECKPOINT_META *ckpt_meta)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *md_cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *internal_session;
+    size_t metadata_value_cfg_len;
+    uint64_t checkpoint_timestamp, current_meta_lsn;
+    char *buf, *metadata_value, *metadata_value_cfg;
+    char ts_string[WT_TS_INT_STRING_SIZE];
+
+    conn = S2C(session);
+
+    buf = NULL;
+    internal_session = NULL;
+    md_cursor = NULL;
+    metadata_value = NULL;
+    metadata_value_cfg = NULL;
+    metadata_value_cfg_len = 0;
+    checkpoint_timestamp = WT_TS_NONE;
+
+    WT_ASSERT_SPINLOCK_OWNED(session, &conn->checkpoint_lock);
+
+    /* We should not pick up a checkpoint with an earlier LSN. */
+    current_meta_lsn =
+      __wt_atomic_load_uint64_acquire(&conn->disaggregated_storage.last_checkpoint_meta_lsn);
+    if (ckpt_meta->metadata_lsn < current_meta_lsn)
+        WT_RET_MSG(session, EINVAL,
+          "Attempting to pick up an older checkpoint: current metadata LSN = %" PRIu64
+          ", new metadata LSN = %" PRIu64,
+          current_meta_lsn, ckpt_meta->metadata_lsn);
+    /*
+     * Warn if we are picking up the same checkpoint again. There's nothing else to do here, goto
+     * err for cleanup.
+     */
+    if (ckpt_meta->metadata_lsn == current_meta_lsn) {
+        __wt_verbose_level(session, WT_VERB_LAYERED, WT_VERBOSE_WARNING,
+          "Picking up the same checkpoint again: metadata LSN = %" PRIu64, ckpt_meta->metadata_lsn);
+        /* Keep previous ret value to avoid overlapping error message */
+        goto err;
+    }
+
+    __wt_verbose_debug1(session, WT_VERB_DISAGGREGATED_STORAGE,
+      "Picking up disaggregated storage checkpoint: metadata_lsn=%" PRIu64,
+      ckpt_meta->metadata_lsn);
+
+    /*
+     * Part 1: Get the metadata of the shared metadata table and insert it into our metadata table.
+     */
+
+    WT_ERR(__disagg_read_meta(session, ckpt_meta, &buf));
+    WT_ERR(__disagg_load_checkpoint_meta(session, buf, &metadata_value, &checkpoint_timestamp));
+
+    __wt_verbose_debug2(session, WT_VERB_DISAGGREGATED_STORAGE,
+      "Picking up disaggregated storage checkpoint: metadata_lsn=%" PRIu64 ", timestamp=%" PRIu64
+      " %s"
+      ", root=\"%s\"",
+      ckpt_meta->metadata_lsn, checkpoint_timestamp,
+      __wt_timestamp_to_string(checkpoint_timestamp, ts_string), metadata_value);
+
+    /* We need an internal session when modifying metadata. */
+    WT_ERR(__wt_open_internal_session(conn, "checkpoint-pick-up", false, 0, 0, &internal_session));
+
+    /* Open up a metadata cursor pointing at our table */
+    WT_ERR(__wt_metadata_cursor(internal_session, &md_cursor));
+
+    WT_ERR(__disagg_save_meta(session, internal_session, md_cursor, metadata_value,
```

Comment:
The open of the internal session -> meta_cursor open -> disagg_save _meta -> disagg_process checkpoint, IMO have a shared responsibility. Which is to all update the metadata information of all tables. 

Can we have a parent function called __disagg_process_metadata which inside does this logic?

Also WDYT about __disagg_apply_shared_meta as a function name?